### PR TITLE
[THREESCALE-2166] Add Upstream Connection policy

### PR DESCRIPTION
**Pull request by davidor**
This PR adds a new policy, "Upstream Connection".

This policy allows to configure several options for the connections to the upstream. In particular, 3 timeouts:  connect_timeout, send_timeout, and read_timeout.

Ref: https://issues.jboss.org/browse/THREESCALE-2166

### CODE REVIEW

**Comment by davidor** [comment]
@vramosp I named the policy "Upstream Connection". If you want to rename it, let me know.

**Comment by vramosp** [comment]
@davidor Would something like 'Upstream Timeout Configuration' be too restrictive?

**Comment by davidor** [comment]
@vramosp In my opinion, that's a bit restrictive because in the future we might want to add other configuration parameters to this policy that are not timeouts.

**Comment by vramosp** [comment]
@davidor I can't think of anything better then.

File: gateway/src/apicast/balancer.lua [hunk 1]
```diff
@@ -52,6 +52,20 @@ local function get_peer(balancer, upstream)
...
+  if not timeouts then return end
+
+  local _, err = balancer:set_timeouts(
+      timeouts.connect_timeout,
+      timeouts.send_timeout,
+      timeouts.read_timeout
+  )
+
+  if err then
+    ngx.log(ngx.WARN, 'Error while setting balancer timeouts: ', err)
```
Comment (eloycoto) [inline gateway/src/apicast/balancer.lua]: I read the resty balance timeout, and the error message is not too descriptive[0]. I would add the connect_timeout, send_timeout and read_timeout in the warning messages. 

https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.lua#L166-L206

File: spec/balancer_spec.lua [hunk 2]
```diff
@@ -23,5 +23,37 @@ describe('apicast.balancer', function()
       assert(apicast_balancer:call({ upstream =  upstream }, balancer))
       assert.spy(balancer.set_current_peer).was.called_with(balancer, '127.0.0.2', 443)
     end)
+
+    it('sets the timeouts in the balancer when they are received in the context', function()
+      local timeouts = {
```
Comment (eloycoto) [inline spec/balancer_spec.lua]: I would add a test with timeouts with negative numbers just in case.

File: gateway/src/apicast/policy/upstream_connection/apicast-policy.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  "configuration": {
+    "type": "object",
+    "properties": {
+      "connect_timeout": {
+        "description": "Timeout for establishing a connection (in seconds).",
+        "type": "integer"
+      },
+      "send_timeout": {
+        "description": "Timeout between two successive write operations (in seconds).",
+        "type": "integer"
```
Comment (eloycoto) [inline gateway/src/apicast/policy/upstream_connection/apicast-policy.json]: Just a note here, could we use a type number with a range? Like this for example: 

https://json-schema.org/understanding-json-schema/reference/numeric.html#range 

So validation for negative numbers can be in the schema.

File: gateway/src/apicast/policy/upstream_connection/apicast-policy.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  "description": "Allows to configure several options for the connections to the upstream",
+  "version": "builtin",
+  "configuration": {
+    "type": "object",
+    "properties": {
+      "connect_timeout": {
+        "description": "Timeout for establishing a connection (in seconds).",
+        "type": "integer"
+      },
+      "send_timeout": {
```
Comment (eloycoto) [inline gateway/src/apicast/policy/upstream_connection/apicast-policy.json]: Just a note here, could we use a type number with a range? Like this for example:

https://json-schema.org/understanding-json-schema/reference/numeric.html#range

So validation for negative numbers can be in the schema.

File: gateway/src/apicast/policy/upstream_connection/apicast-policy.json (hunk 3 above)
Comment (davidor) [reply gateway/src/apicast/policy/upstream_connection/apicast-policy.json]: Good idea. Added :+1:

File: gateway/src/apicast/balancer.lua (hunk 1 above)
Comment (davidor) [reply gateway/src/apicast/balancer.lua]: You're right, the error does not show the values. I added an `inspect()` there :+1:
